python/dgl/nn/pytorch/conv/edgegraphconv.py

```diff
@@ -0,0 +1,289 @@
+
+            Case 2:
+            External weight is provided while at the same time the module
+            has defined its own weight parameter.
+
+        Note
+        ----
+        * Input shape: :math:`(N, *, \text{in_feats})` where * means any number of additional
+          dimensions, :math:`N` is the number of nodes.
+        * Output shape: :math:`(N, *, \text{out_feats}) * \text{edge_weight_dimension}` where all
+          but the last dimension are the same shape as the input.
+        * Weight shape: :math:`(\text{in_feats}, \text{out_feats})`.
+        * Edge weight shape: :math:`(\text{num_edges}, \text{edge_weight_dimension})`.
+        """
+        with graph.local_scope():
+            if not self._allow_zero_in_degree:
+                if (graph.in_degrees() == 0).any():
+                    raise DGLError('There are 0-in-degree nodes in the graph, '
+                                   'output for those nodes will be invalid. '
+                                   'This is harmful for some applications, '
+                                   'causing silent performance regression. '
+                                   'Adding self-loop on the input graph by '
+                                   'calling `g = dgl.add_self_loop(g)` will resolve '
+                                   'the issue. Setting ``allow_zero_in_degree`` '
+                                   'to be `True` when constructing this module will '
+                                   'suppress the check and let the code run.')
+            if edge_weight is None:
+                raise DGLError("Edge weight is required for this layer type (EdgeGraphConv)")
+            assert edge_weight.shape[0] == graph.number_of_edges()
+            aggregate_fn = fn.u_mul_e('h', '_edge_weight', 'm')
+
+            # (BarclayII) For RGCN on heterogeneous graphs we need to support GCN on bipartite.
+            feat_src, feat_dst = expand_as_pair(feat, graph)
+            if self._norm in ['left', 'both']:
+                degs = graph.out_degrees().float().clamp(min=1)
+                if self._norm == 'both':
+                    norm = th.pow(degs, -0.5)
+                else:
+                    norm = 1.0 / degs
+                shp = norm.shape + (1,) * (feat_src.dim() - 1)
+                norm = th.reshape(norm, shp)
+                feat_src = feat_src * norm
+            graph.srcdata['h'] = feat_src
+
+            if weight is not None:
+                if self.weight is not None:
+                    raise DGLError('External weight is provided while at the same time the'
+                                   ' module has defined its own weight parameter. Please'
+                                   ' create the module with flag weight=False.')
+            else:
+                weight = self.weight
+
+            single_channel_rsts = []
+            for i in range(edge_weight.shape[1]):
```

Comment:
I think I've done what you asked -- if that's not the case, can you guide me in the right direction?

Reply:
The question is how different is it from `GraphConv` with `edge_weight` specified [here](https://github.com/dmlc/dgl/blob/master/python/dgl/nn/pytorch/conv/graphconv.py#L337)? Also, what's the purpose of `th.stack([edge_weight] * self._in_feats, 2)`?

Reply:
It's the same as GraphConv but it supports a multi-dimensional edge weight. GraphConv currently supports one dimension only. Including this as part of GraphConv is a reasonable approach. 

The purpose of `th.stack([edge_weight] * self._in_feats, 2)` is to avoid a dimension mismatch during the aggregation step. The node features are self._in_feats dimensional and edge features are edge_weight_dimension dimensional. fn.u_mul_e can't multiply these directly.

Reply:
Something like below should work.

```python
import dgl
import dgl.function as fn
import torch

g = dgl.rand_graph(num_nodes=5, num_edges=20)
in_feats = 5
num_ew_channels = 3
h = torch.randn(g.num_nodes(), in_feats)
eweight = torch.randn(g.num_edges(), num_ew_channels, 1)
g.ndata['h'] = h
g.edata['w'] = eweight
g.update_all(fn.u_mul_e('h', 'w', 'm'), fn.sum('m', 'h_new'))
```

`g.ndata['h_new']` is of shape `(N, H, F)`, where `N` is the number of nodes, `H` is the number of eweight channels and `F` is the feature size.